sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -56,6 +56,210 @@ that is compiled from several separate ``.less`` files in the BokehJS source
+test suites with ``node make test:suite_name``. Known tests suites are:
+
+* ``node make test:size``
+* ``node make test:defaults``
+* ``node make test:unit``
+* ``node make test:integration``
+
+The last to can be run with ``node make test:lib``. Unit and integration tests are
+run in a web browser (see requirements), which is started automatically with the
+right settings to guarantee consistent test results.
+
+To review visual tests' output, start bokehjs' devtools server:
+
+.. code-block:: sh
+
+    $ node test/devtools server
+    listening on 127.0.0.1:5777
+
+and navigate to ``/integration/report``. Devtools server can be also used to
+manually inspect and debug tests. For that the following endpoints are available:
+
+* ``/unit``
+* ``/defaults``
+* ``/integration``
+
+Those load bokehjs and tests, but don't anything. You have to issue ``Tests.run_all()``
+in JavaScript console. This allows to you to set breakpoints before running code. You
+can filter out tests by providing a string keyword or a regular expression. Alternatively
+you can run tests immediately with these endpoints:
+
+* ``/unit/run``
+* ``/defaults/run``
+* ``/integration/run``
+
+You can use ``?k=some%20text`` to filter tests by a keyword.
+
+CI and visual testing
+~~~~~~~~~~~~~~~~~~~~~
+
+``test:integration`` does two types of tests:
+
+* textual baseline tests
+* visual/screenshot tests
+
+Textual baselines are cross-platform compatible and can be generated locally (on
+supported platforms) or in CI. Visual testing is platform depended and fairly
+sensitive to system configuration (especially in regard to differences in font
+rendering). Visual tests can be performed locally, but given that baseline images
+for all three supported platforms have to be updated, the preferred approach is
+to generate images and compare them in CI.
+
+The full procedure for visual testing is as follows:
+
+1. Make changes to the repository and write new tests or update existing.
+2. Use ``node make tests`` to incrementally test your changes on your system.
+3. Commit changes to textual baselines (``test/baselines/*``).
+4. Push your changes to GitHub and wait for CI to finish.
+5. If you added new tests and CI will expectedly fail with "missing baseline
+   images" error message.
+6. If tests passed then your are done.
+7. If tests failed, go to bokehjs' GitHub_Actions_ page. Find the most recent
+   test run for your PR and download the associated ``bokehjs-report`` artifact.
+8. Unzip the artifact archive.
+9. Assuming devtools server is running in the background, go to ``/integration/report?platform=name``
+   where ``name`` is either ``linux``, ``macos`` or ``windows`` and review test output
+   for each platform. If there are no unintentional differences, then commit all
+   new or modified files under ``test/baselines/{linux,macos,windows}``.
+10. Push your changes again to GitHub and verify that tests pass this time.
```

Comment:
This is all so complicated. For there to be any hope at all that anyone will ever try to follow through with this, there would need to be annotated screenshots showing the steps along the way to provide people with much more anchoring context than a wall of text. Ideally, there would be a step by step video as well. The is simply not enough (not nearly enough) to get a brand new person going from zero.

Other things that could help:

* constructing exact commands to needed download/update things locally, and posting them back to as a PR comment so that they can be cut/pasted. 
* split things in to "common case process" and "full process" to avoid giving the impression that all this will be required for any change at all (what it seems like now)
* Automate some tasks, so that there aren't caveats like "Assuming XZY is running" (already in the middle of ten other steps)

Reply:
This is just the initial wording and the procedure isn't fully described yet. I need to describe how to deal with tests that result in pixel fluctuations (which I believe at this point to be a bug in chromium). I've been working on a script to make this simpler, but it looks like the new APIs around GitHub actions and especially artifact management aren't reliable (if not plainly broken; I need to submit a bug to GitHub regarding this). However, even when I get the script working, as my current understanding goes, the token required to allow access to artifacts via the API, requires access scope that only repository members can get. So external contributors and especially first time contributors will have go through the manual procedure anyway. The whole affair with this new approach to bokehjs testing will have to refined over time. My intention is to merge this as-is, well with a few tweaks first, and gradually refine this.